### PR TITLE
Markygriff/protobuf packaging

### PR DESCRIPTION
**Pull request by markygriff**

### CODE REVIEW

File: CAN-D/Src/gps.c [hunk 1]
```diff
@@ -70,31 +71,42 @@ void APP_GPS_BufferGPSString(char* dataString, size_t dataLength)
...
+            gpsRxMsg = event.value.p;
             // TODO: BO: Moving around configurations, fix this
             //if (mAppConfiguration.SDStorage == APP_ENABLE) {
             if (1) {
                 // Write data to SD card
-                APP_FATFS_LogSD((const uint8_t*)data->raw, 128, gpsLogIdentifier);
+                APP_FATFS_LogSD((const uint8_t*)gpsRxMsg->raw, GPS_SD_DATA_SZ_BYTES, gpsLogIdentifier);
             }
 
+            fromEmbeddedMsg.contents.gpsDataChunk.size = 128;
```
Comment (bofrim) [inline CAN-D/Src/gps.c]: Magic number. Could it be replaced with: `GPS_SD_DATA_SZ_BYTES `

File: CAN-D/Src/gps.c [hunk 2]
```diff
@@ -70,31 +71,42 @@ void APP_GPS_BufferGPSString(char* dataString, size_t dataLength)
...
+            fromEmbeddedMsg.which_contents = 2;
```
Comment (bofrim) [inline CAN-D/Src/gps.c]: Should be: `FromEmbedded_gpsDataChunk_tag`

File: CAN-D/Src/can.c [hunk 3]
```diff
@@ -231,31 +232,46 @@ void APP_CAN_MonitorTask(void const* argument)
...
                 // Write data to SD card
-                APP_FATFS_LogSD((const uint8_t*)msg->data, 8, canLogIdentifier);
+                APP_FATFS_LogSD((const uint8_t*)canRxMsg->data, CAN_RX_MSG_DATA_SZ_BYTES, canLogIdentifier);
             }
 
+            // Pack the protobuf message
+            fromEmbeddedMsg.contents.canDataChunk.data.size = 8;
+            fromEmbeddedMsg.contents.canDataChunk.has_id = 1;
+            fromEmbeddedMsg.contents.canDataChunk.has_data = 1;
+            fromEmbeddedMsg.which_contents = 1;
```
Comment (bofrim) [inline CAN-D/Src/can.c]: Should be `FromEmbedded_canDataChunk_tag`

File: CAN-D/Src/can.c [hunk 4]
```diff
@@ -231,31 +232,46 @@ void APP_CAN_MonitorTask(void const* argument)
...
 
         // Pend on any CAN Rx data
         event = osMessageGet(CANRxQueueHandle, 0);
         if (event.status == osEventMessage) {
-            msg = event.value.p;
+            canRxMsg = event.value.p;
             if (mAppConfiguration.SDStorage == APP_ENABLE) {
                 // Write data to SD card
-                APP_FATFS_LogSD((const uint8_t*)msg->data, 8, canLogIdentifier);
+                APP_FATFS_LogSD((const uint8_t*)canRxMsg->data, CAN_RX_MSG_DATA_SZ_BYTES, canLogIdentifier);
```
Comment (bofrim) [inline CAN-D/Src/can.c]: Maybe `CAN_RX_MSG_DATA_SZ_BYTES`

File: CAN-D/Src/gps.c [hunk 5]
```diff
@@ -70,31 +71,42 @@ void APP_GPS_BufferGPSString(char* dataString, size_t dataLength)
   */
 void APP_GPS_MonitorTask(void const* argument)
 {
-    uint8_t usbTxCnt = 0;
+    GPSData gpsRx = { 0 };
+    volatile uint8_t usbTxCnt = 0;
     osEvent event;
-    GPSData* data;
+    GPSData* gpsRxMsg = &gpsRx;
+    size_t usbMaxMsgLen = GPS_USB_DATA_SZ_BYTES + 10; // Max length of the serialized data
```
Comment (bofrim) [inline CAN-D/Src/gps.c]: Maybe we could have a `PROTO_BUFFER_OVERHEAD` defined in the proto handler header

File: CAN-D/Src/can.c [hunk 6]
```diff
@@ -231,31 +232,46 @@ void APP_CAN_MonitorTask(void const* argument)
...
+            fromEmbeddedMsg.contents.canDataChunk.has_data = 1;
+            fromEmbeddedMsg.which_contents = 1;
+
+            memcpy(fromEmbeddedMsg.contents.canDataChunk.data.bytes, canRxMsg->data, CAN_RX_MSG_DATA_SZ_BYTES);
+            fromEmbeddedMsg.contents.canDataChunk.id = (canRxMsg->header->StdId & CAN_RX_MSG_STDID_MASK);
+            usbTxNumBytes = APP_PROTO_HANDLE_bufferFromEmbeddedMsg(&fromEmbeddedMsg, (uint8_t*)usbTxMsg, usbMaxMsgLen);
+
             usbTxCnt = 0;
-            while (APP_USB_Transmit((uint8_t*)msg->data, CAN_USB_DATA_SZ_BYTES) == 1) {
+            while (APP_USB_Transmit((uint8_t*)usbTxMsg, usbTxNumBytes) == 1) {
```
Comment (bofrim) [inline CAN-D/Src/can.c]: `while( ... ==USBD_OK)`

**Comment by bofrim** [comment]
and a conflict

File: CAN-D/Src/can.c (hunk 3 above)
Comment (markygriff) [reply CAN-D/Src/can.c]: Thanks! Couldn't find this one.

**Pull request merged**